[special] Add `i1` and `i1e`

kshitij12345 opened the pull request:
Reference: #50345 

* [x] Check Docs https://12721710-65600975-gh.circle-artifacts.com/0/docs/special.html
* [x] Investigate fp32 failure on CI?! (Fails on clang. Reproduced locally with clang-11)
* [ ] Kernel vs Composite?
* [x] Autograd for `i0e` for zero?

kshitij12345 committed:
init support for i0e

kshitij12345 committed:
Merge branch 'master' into develop/special/i0e

kshitij12345 committed:
add overrides entry

kshitij12345 committed:
undo unnecessary change

kshitij12345 committed:
update doc example

kshitij12345 committed:
make flake8 happy

kshitij12345 committed:
Merge branch 'master' into develop/special/i0e

kshitij12345 committed:
add test to check for min, max, smallest value for dtype

kshitij12345 committed:
add entry to special.rst

kshitij12345 committed:
docs: fix signature typo

kshitij12345 committed:
update docs and comment

kshitij12345 committed:
Merge branch 'master' into develop/special/i0e

kshitij12345 committed:
Merge branch 'master' into develop/special/i0e

kshitij12345 committed:
address review

kshitij12345 committed:
Merge branch 'master' of https://github.com/pytorch/pytorch into develop/special/i0e

kshitij12345 committed:
try fixing docs

kshitij12345 committed:
use structured kernel

kshitij12345 committed:
Merge branch 'master' into develop/special/i0e

kshitij12345 committed:
retrigger CI

kshitij12345 committed:
Merge branch 'master' into develop/special/i0e

kshitij12345 committed:
support i1, i1e and autograd for i0,i0e,i1,i1e

kshitij12345 committed:
use auto keyword on lhs of assignment

kshitij12345 committed:
cuda support for i1 and i1e

kshitij12345 committed:
update derivatives and sample

kshitij12345 committed:
Merge branch 'master' into develop/special/i1-i1e

kshitij12345 committed:
fix merge artifacts

kshitij12345 committed:
update wrapper name

kshitij12345 committed:
Merge branch 'master' into develop/special/i1-i1e

kshitij12345 committed:
fix merge artifacts

kshitij12345 committed:
update entry

kshitij12345 committed:
make mypy happy

kshitij12345 committed:
update at::where usage

kshitij12345 commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -3252,6 +3252,28 @@ std::tuple<Tensor, Tensor> polar_backward(
+      at::kHalf, at::kBFloat16, self.scalar_type(), "i1_backward", [&]() {
+        auto eps = std::numeric_limits<scalar_t>::epsilon();
+        auto self_is_not_tiny = self.abs() > eps;
+        auto safe_self = at::where(self_is_not_tiny, self, at::full({}, eps, self.options()));
```
(Note to self): Following used to work. (investigate)
auto safe_self = at::where(self_is_not_tiny, self, eps);

kshitij12345 committed:
relaxation for test

kshitij12345 committed:
add docs

kshitij12345 committed:
fix linter issues

kshitij12345 committed:
Merge branch 'master' into develop/special/i1-i1e

kshitij12345 commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -1649,6 +1649,23 @@ def wrapped_fn(x):
     return wrapped_fn
 
 
+def np_unary_ufunc_integer_promotion_wrapper_with_astype(fn):
```
Better Name?

mruberry replied:
What about changing the previous function to use `astype` instead of `dtype=` instead of creating this new function?

kshitij12345 replied:
With `astype`, the computation still happens in double but we just cast the result to `float` in our case. 
With `dtype`, the computation actually happens in the mentioned dtype as well, so I believe it makes sense to use `dtype` wherever applicable.

What do you think?

mruberry replied:
I meant what if we use astype() BEFORE calling the operation?

kshitij12345 commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -3631,7 +3668,27 @@ def gradcheck_wrapper_triangular_input(op, input, *args, upper=False, **kwargs):
+    UnaryUfuncInfo('special.i1',
+                   aten_name='special_i1',
+                   ref=np_unary_ufunc_integer_promotion_wrapper_with_astype(scipy.special.i1),
+                   decorators=(precisionOverride({torch.bfloat16: 2.0,
```
Add more details

kshitij12345 committed:
update c++ out api

kshitij12345 committed:
update license info

kshitij12345 committed:
add test for c++ out variants

kshitij12345 commented on torch/csrc/api/include/torch/special.h:
```diff
@@ -161,8 +161,40 @@ inline Tensor i0e(const Tensor& self) {
 
-inline Tensor i0e_out(const Tensor& self) {
-  return torch::special_i0e(self);
+inline Tensor i0e_out(Tensor& result, const Tensor& self) {
```
Missed out on this in the previous PR.

Have added a very basic test in `special.cpp`

mruberry replied:
Would you move this fix to a separate PR? I don't know if we need a test for this

kshitij12345 replied:
Have kept the fix. But removed the test. Is it okay?

mruberry replied:
Sounds good ;)

kshitij12345 committed:
add overrides entry

kshitij12345 committed:
qualify type ignore with error code

ezyang commented:
these are structured so i am happy, leaving rest of review to others

kshitij12345 committed:
relaxation for i1 and float32

kshitij12345 committed:
fix missing bracket

mruberry commented:
Hey @kshitij12345, hope you're doing well. Just checking-in: anything I can do to help here?

kshitij12345 commented:
> Hey @kshitij12345, hope you're doing well. Just checking-in: anything I can do to help here?

I have few doubts. Will connect with you once I have enlisted them. Thanks.

Feel free to have a cursory look at the PR.

Thanks! :)

kshitij12345 commented:
@mruberry 

The discrepancy between clang and gcc is because of how things are vectorized.
With vectorization and optimization flags on, gcc produces a different result which luckily is closer to the expected value.
However with optimization of gcc agrees with clang. So I think we have to relax the tolerance.

Note: clang's results are consistent with optimization flags on and off.

Godbolt Reproduction: https://godbolt.org/z/o539d11Tq

Another things I tried (as sanity check) was to cast the float to double for computation (similar to scipy), in which case both clang and gcc tests pass.

mruberry commented:
> @mruberry
> 
> The discrepancy between clang and gcc is because of how things are vectorized.
> With vectorization and optimization flags on, gcc produces a different result which luckily is closer to the expected value.
> However with optimization of gcc agrees with clang. So I think we have to relax the tolerance.
> 
> Note: clang's results are consistent with optimization flags on and off.
> 
> Godbolt Reproduction: https://godbolt.org/z/o539d11Tq
> 
> Another things I tried (as sanity check) was to cast the float to double for computation (similar to scipy), in which case both clang and gcc tests pass.

That's really interesting. How wide does the tolerance need to be to pass on both clang and gcc with float32? Can we only apply the wider tolerance when testing against clang?

kshitij12345 commented:
> That's really interesting. How wide does the tolerance need to be to pass on both clang and gcc with float32?

With the current approximation (double precision) coefficients, the tolerance required is `2e-4`. With the approximation coefficients for single precision (cephes single-precision), IIRC it can be brought down to `1e-4` plus it uses lower order chebyshev polynomial for approximation, thus should be faster (will run the benchmark and update).

**Note**: IIRC, with single-precision approximation, both gcc and clang produced consistent output irrespective of the vectorization flags.

> Can we only apply the wider tolerance when testing against clang?

I think we can manage to have a flag for which compiler was the lib build with (you had shared a reference previously I think). However I don't think that is a good idea because the root cause is not tied to the compiler itself but the passed flags.
* clang consistently computes the correct values (for single-precision) (irrespective of optimization)
* GCC values change as per optimization. (So overtime if a new CI build pops up with GCC debug or something then it will start failing as well)


Thanks!

kshitij12345 commented:
@mruberry your guess was correct. First derivative of `i0e` looks to be discontinuous at `0` and hence the trouble. Thanks!

kshitij12345 committed:
changes

* remove support for bfloat16 and half
* support single-precision approximation

kshitij12345 committed:
update relevant tests

kshitij12345 commented:
Benchmarks for approximation with single-precision coefficients.

**Note**: std-Array is current implementation just wrapped in std::array for convenience.

```
[--------- i0 : input dtype torch.float32 device cpu ---------]
                         |  single-precision coeff  |  std-Array
1 threads: -----------------------------------------------------
      (128,)             |              7.9         |        8.2
      (256, 128)         |           1596.9         |     1622.4
      (2, 512, 256)      |          12739.0         |    13039.8
      (2, 64, 256, 128)  |         205432.9         |   208691.4

Times are in microseconds (us).

```

kshitij12345 committed:
Merge branch 'master' into develop/special/i1-i1e

kshitij12345 committed:
add comment for i0e

kshitij12345 committed:
remove trailing space

kshitij12345 committed:
revert from using std::array to c-style array :(

kshitij12345 committed:
update supported_backward for i0 and i0e

kshitij12345 commented:
@mruberry please review once :). The major things are in-place.

mruberry commented on aten/src/ATen/native/Math.h:
```diff
@@ -1279,15 +1382,106 @@ calc_i0e(T _x) {
+  if (x <= 8.0) {
+    const auto A = chebyshev_coefficients_i1e_A<T>();
+    T y = (x / 2.0) - 2.0;
+    const T out = static_cast<T>(std::exp(x) * x * chbevl(y, A, 17));
```
If these values are the only two differences, maybe create a common function that accepts these as arguments? Then a template can specialize float and double calling the common helper with the desired values

kshitij12345 replied:
I tried using `std::array` but mobile and mac builds were failing.
I was planning to do some refactoring later. For now, I think we can return coefficients and their length as tuple.

kshitij12345 replied:
Done with `std::tuple` for now.

mruberry commented on aten/src/ATen/native/Math.h:
```diff
@@ -1279,15 +1382,106 @@ calc_i0e(T _x) {
+ */
+template <typename T>
+static inline typename std::enable_if<std::is_floating_point<T>::value, T>::type
+calc_i1e(T _x) {
```
Same strategy with a common function should work here, too?

mruberry commented on aten/src/ATen/native/cuda/Math.cuh:
```diff
@@ -299,14 +407,100 @@ static inline C10_HOST_DEVICE scalar_t calc_i0e(scalar_t _x) {
   return static_cast<scalar_t>(chbevl(static_cast<accscalar_t>(32.0 / x - 2.0), B, 25) / ::sqrt(x));
 }
 
+template <typename scalar_t>
```
These functions are identical to their CPU counterparts, right?

kshitij12345 replied:
Even with upcasting for the computation and then downcasting to original dtype has issue (mostly due to the way inherent characteristics of these dtypes). 
I think it would be better to support them if there is demand for the same.
```python
import torch
import scipy.special

t = torch.tensor(8.3750, dtype=torch.bfloat16)
print("TORCH:", torch.special.i1(t))
print("SCIPY:", scipy.special.i1(t.to(torch.float).numpy()))
print("BFLOAT16 CONVERSION:", torch.tensor(570.0086, dtype=torch.bfloat16))

expected = torch.special.i1(t.to(torch.double))
actual = torch.special.i1(t)

abs_err = torch.abs(expected - actual)
print("ABSOLUTE ERROR:", abs_err)
print("RELATIVE ERROR:", abs_err/expected)
```
Out:
```
TORCH: tensor(572., dtype=torch.bfloat16)
SCIPY: 570.0086
BFLOAT16 CONVERSION: tensor(572., dtype=torch.bfloat16)
ABSOLUTE ERROR: tensor(1.9914, dtype=torch.float64)
RELATIVE ERROR: tensor(0.0035, dtype=torch.float64)
```

mruberry commented on aten/src/ATen/native/cuda/Math.cuh:
```diff
@@ -299,14 +407,100 @@ static inline C10_HOST_DEVICE scalar_t calc_i0e(scalar_t _x) {
+template <>
+inline C10_HOST_DEVICE float calc_i1(float _x) {
+  using scalar_t = float;
+  using accscalar_t = at::acc_type<scalar_t, true>;
```
The accumulate type of float is just float, right?

mruberry commented on aten/src/ATen/native/cuda/Math.cuh:
```diff
@@ -299,14 +407,100 @@ static inline C10_HOST_DEVICE scalar_t calc_i0e(scalar_t _x) {
 }
 
+template <typename scalar_t>
+static inline C10_HOST_DEVICE scalar_t calc_i1(scalar_t _x) {
```
So bfloat16 and float16 would call this function, and then be upcast to float, but this is not using the float-specific values?

kshitij12345 replied:
Killed this part of the code. Kernel is note instantiated for float16 and bfloat16.

mruberry commented on test/cpp/api/special.cpp:
```diff
@@ -13,3 +13,35 @@ TEST(SpecialTest, special) {
     torch::special::gammaln(t);
 }
+
+template <typename Op, typename OpOut>
```
What's up with this new test? Is this maybe more appropriate for a separate PR?

kshitij12345 replied:
Removed.

mruberry commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -3340,6 +3340,28 @@ std::tuple<Tensor, Tensor> polar_backward(
+Tensor i1_backward(const Tensor& grad, const Tensor& self, const Tensor& result) {
+  return AT_DISPATCH_FLOATING_TYPES_AND2(
+      at::kHalf, at::kBFloat16, self.scalar_type(), "i1_backward", [&]() {
+        auto eps = std::numeric_limits<scalar_t>::epsilon();
```
This is going to need a comment explaining the approach

kshitij12345 replied:
Updated.

mruberry commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -3340,6 +3340,28 @@ std::tuple<Tensor, Tensor> polar_backward(
+      at::kHalf, at::kBFloat16, self.scalar_type(), "i1_backward", [&]() {
+        auto eps = std::numeric_limits<scalar_t>::epsilon();
+        auto self_is_not_tiny = self.abs() > eps;
+        auto safe_self = at::where(self_is_not_tiny, self, at::full({}, eps, self.options()));
```
What does at:full do when called with an empty initializer list?

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -1888,6 +1888,23 @@ def wrapped_fn(x):
+        return dtype in [np.bool_, bool, np.uint8, np.int8, np.int16, np.int32, np.int64]
+
+    # NOTE: Promotion in PyTorch is from integer types to the default dtype
+    np_dtype = torch_to_numpy_dtype_dict[torch.get_default_dtype()]
```
Because the default dtype can change, the acquisition of the np_dtype should happen when the function is called, not when it's wrapped

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -4245,19 +4285,43 @@ def gradcheck_wrapper_triangular_input(op, input, *args, upper=False, **kwargs):
                    decorators=(precisionOverride({torch.bfloat16: 3e-1,
                                                   torch.float16: 5e-1}),),
+                   skips=(
+                       SkipInfo('TestOpInfo', 'test_supported_backward',
```
We now have new metadata:

https://github.com/pytorch/pytorch/blob/1461859fdebe1346accf50751854b5448b832be8/torch/testing/_internal/common_methods_invocations.py#L161

which avoids these skips

kshitij12345 replied:
Note: 
Would be nice we if could just pass `backward_dtypes` for UnaryUfunc,
However with the way, `None` is handled for `backward_dtypesIf*`, for using it with `UnaryUfunc` we have to specify all of them as UnaryUfunc specifies all `dtypesIf*`.
https://github.com/pytorch/pytorch/blob/1fc89d9ffcea234ccf9a35f60a7afc3cb0d406ee/torch/testing/_internal/common_methods_invocations.py#L210-L216

https://github.com/pytorch/pytorch/blob/1fc89d9ffcea234ccf9a35f60a7afc3cb0d406ee/torch/testing/_internal/common_methods_invocations.py#L385-L387

mruberry replied:
Yes but we should fix that so it has the same logic as a regular OpInfo

kshitij12345 replied:
Noted :)

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -4245,19 +4285,43 @@ def gradcheck_wrapper_triangular_input(op, input, *args, upper=False, **kwargs):
+                   ref=np_unary_ufunc_integer_promotion_wrapper_with_astype(scipy.special.i1),
+                   decorators=(precisionOverride({torch.float: 1e-4}),),
+                   dtypes=all_types_and(torch.bool),
+                   dtypesIfCPU=all_types_and(torch.bool),
```
For types, earlier it looked like there was vectorized support for bfloat16 and half, and on CUDA there's accommodation to promote to the accumulate type in the kernel, but `all_types` does not include half or bfloat16 -- should the half and bfloat16 code just be removed (for now)? Alternatively support for it should be added and tested

kshitij12345 replied:
Killed the support for `bfloat` and `half`. I think it would be better to add support for them when requested. Otherwise it would be a bit messy, 
* Even with upcasting for computation, the result won't be very much correct. Refer https://github.com/pytorch/pytorch/pull/56352#discussion_r626111962
* `i1 and i1e` CPU won't support half or bfloat16
* Since `i0 and i0e` depends on `i1 and i1e` for backward, `backward_dtypesIfCUDA` for `i0 and i0e` will be `torch.half, torch.bfloat16, torch.float32, torch.float64` while for `CPU` it will be`torch.float32, torch.float64`

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -2964,6 +2981,29 @@ def sample_inputs_entr(op_info, device, dtype, requires_grad, **kwargs):
+                                       requires_grad=requires_grad)))
+
+    if dtype.is_floating_point:
+        if op_info.op != torch.special.i0e:
```
code nit: two if statements in a row (combine them)

kshitij12345 replied:
Updated. Thanks!

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -2964,6 +2981,29 @@ def sample_inputs_entr(op_info, device, dtype, requires_grad, **kwargs):
+
+    if dtype.is_floating_point:
+        if op_info.op != torch.special.i0e:
+            # Second-order derivative for `i0e` at `0`
```
This comment suggests the code applies to i0e, but the if statement applies to everything BUT i0e

kshitij12345 replied:
Have updated the comment, hopefully it is clearer now.

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -2964,6 +2981,29 @@ def sample_inputs_entr(op_info, device, dtype, requires_grad, **kwargs):
+                            requires_grad=requires_grad)
+
+            with torch.no_grad():
+                t[torch.randn_like(t) > 0] = 0
```
What's going on here? This is randomly adding zeros to t? But isn't this supposed to be avoiding zeros?

kshitij12345 replied:
Have updated the comment, hopefully it is clearer now.

mruberry reviewed (commented):
Hey @kshitij12345, this looks really good overall; @anjali411 and I got to take a look and we made some comments inline, we're a little confused in a few places, have a few suggestions for code cleanup, and recommend separating C++ test update (or maybe not doing it?)

kshitij12345 committed:
remove out_api test

kshitij12345 committed:
add backward_dtypes entry

kshitij12345 committed:
refactor kernel code

kshitij12345 committed:
update comment

kshitij12345 committed:
add comment regarding the sample at 0

kshitij12345 committed:
update samples input for i0

kshitij12345 committed:
update comment for derivative

kshitij12345 committed:
remove stray files

kshitij12345 committed:
remove vec256_bfloat16 dead-code

kshitij12345 commented:
@mruberry @anjali411 Thank you for the quick review :)

Have mostly addressed the changes
* Slightly refactored the computation code (so now we don't specialize the template for float)
* Killed the code for `float16` and `bfloat16`
* Added comment in `derivatives.yaml`
* Used `backward_dtypes` in OpInfo entry.
* Added a comment regarding testing the gradient at `0` in sample_inputs_i0_i1.
* Updated the acquiring of default dtype in the function call instead of wrapping of function.

Should be ready for another round.

Thanks again!!

mruberry commented:
ASAN test failure looks real: 

```
May 05 22:33:21   test_reference_numerics_hard_i0_cpu_bfloat16 (__main__.TestUnaryUfuncsCPU) ... /var/lib/jenkins/workspace/aten/src/ATen/native/Math.h:1369:10: runtime error: 2.50481e+215 is outside the range of representable values of type 'float'
```

what's going on there?

kshitij12345 commented:
:( Ah I was hopeful that this would work. Will do the clean-up in a seperate PR. Thanks for the heads-up :)

kshitij12345 commented:
Have reset back to the working commit. The `asan` failure on this one looks unrelated :)

kshitij12345 committed:
retrigger CI

kshitij12345 commented:
@mruberry Gentle ping :)

kshitij12345 committed:
Merge branch 'master' into develop/special/i1-i1e

kshitij12345 commented on torch/csrc/autograd/FunctionsManual.h:
```diff
@@ -230,6 +230,14 @@ std::tuple<Tensor, Tensor> householder_product_backward(const Tensor& grad, cons
+    const Tensor& self,
+    const Tensor& result);
 Tensor lu_unpack_backward(
   const std::vector<torch::autograd::Variable>& grads,
```
Is it ok to make this `const vectorlist& grads` in this PR or leave it as is?

https://github.com/pytorch/pytorch/blob/c371542efc31b1abfe6f388042aa3ab0cef935f2/torch/csrc/autograd/function_hook.h#L12

kshitij12345 committed:
Merge branch 'master' into develop/special/i1-i1e

mruberry commented on aten/src/ATen/cpu/vec256/vec256_base.h:
```diff
@@ -419,6 +419,12 @@ struct Vec256 {
     return map(calc_i0e);
   }
+  Vec256<T> i1() const {
+    return map(calc_i1);
```
If the vectorized implementations are just mapping the function, is there a reason to implement them at all? We support CPU kernels without vectorized implementations

kshitij12345 replied:
I'll benchmark and remove the psuedo-vectorized implementations if they are not worth it. Thanks!

kshitij12345 replied:
Quick benchmark shows that there is no real gain with `map`. Will remove the vectorized path.

```
[------ i1 : input dtype torch.float32 device cpu ------]
                         |  non-vectorized  |  vectorized
1 threads: ----------------------------------------------
      (128,)             |         4.1      |       4.2  
      (256, 128)         |       679.9      |     706.4  
      (2, 512, 256)      |      5499.7      |    5723.6  
      (2, 64, 256, 128)  |     90534.8      |   89307.4  

Times are in microseconds (us).

[------ i1 : input dtype torch.float64 device cpu ------]
                         |  non-vectorized  |  vectorized
1 threads: ----------------------------------------------
      (128,)             |          7.9     |        8.0 
      (256, 128)         |       1735.0     |     1712.8 
      (2, 512, 256)      |      13505.5     |    13648.6 
      (2, 64, 256, 128)  |     227570.0     |   229265.9 

Times are in microseconds (us).

```

<details>

<summary> Benchmark Code </summary>

```python
import torch
import itertools
import time
from torch.utils.benchmark import Timer
from torch.utils.benchmark import Compare
import sys
import pickle

print('Using pytorch %s' % (torch.__version__))

cuda_shapes = [(128,), (256, 128), (2, 512, 256), (2, 64, 256, 128), (4, 2, 512, 256, 128)]
cpu_shapes = [(128,), (256, 128), (2, 512, 256), (2, 64, 256, 128)]
cpu_dtypes = [torch.float, torch.double] # torch.bfloat16, 
cuda_dtypes = [torch.bfloat16, torch.float16, torch.float, torch.double]
results = []
repeats = 10

for device in ['cpu']:
    if device=='cpu':
        dtypes = cpu_dtypes
        shapes = cpu_shapes
    else:
        dtypes = cuda_dtypes
        shapes = cuda_shapes

    for dtype in dtypes:
        for shape in shapes:
            t = torch.randn(shape, device=device, dtype=dtype)
            
            tasks = [("torch.special.i1(t)", "vectorized", "")]
            timers = [Timer(stmt=stmt, label=f"i1 : input dtype {dtype} device {device}", sub_label=f"{(shape)}", description=desc, globals=globals()) for stmt, desc, label in tasks]

            for i, timer in enumerate(timers * repeats):
                results.append(
                    timer.blocked_autorange()
                )
                print(f"\r{i + 1} / {len(timers) * repeats}", end="")
                sys.stdout.flush()

with open('bessel-compute-vectorized.pkl', 'wb') as f:
    pickle.dump(results, f)

comparison = Compare(results)
comparison.print()

```

</details>

mruberry replied:
I knew you'd actually run the benchmarks ;) Nice follow-up, @kshitij12345

Looks like the difference is in the noise

mruberry commented on test/test_unary_ufuncs.py:
```diff
@@ -1186,34 +1186,46 @@ def test_i0_special(self, device, dtype):
     @unittest.skipIf(not TEST_SCIPY, "SciPy not found")
-    def test_special_i0e_vs_scipy(self, device, dtype):
-        def check_equal(t):
+    def test_special_i0_i1_vs_scipy(self, device, dtype):
```
How would we need to change the reference numerics tests so that we can delete this operation?

kshitij12345 replied:
Near the following line
https://github.com/pytorch/pytorch/blob/71f4c5c1f436258adc303b710efb3f41b2d50c4e/test/test_unary_ufuncs.py#L183 

 we can probably do,
```python
info = torch.finfo(dtype)
vals = _float_extremals  + [info.min, info.max, info.eps, info.tiny]
```

However we will still miss out on the `range` test
https://github.com/pytorch/pytorch/blob/71f4c5c1f436258adc303b710efb3f41b2d50c4e/test/test_unary_ufuncs.py#L1209-L1214

can we add this range based test as an opt-in test for unary-ufuncs?

mruberry replied:
I think I'm missing something here: should we be randomly generating values within that range as part of test_reference_numerics_hard?

mruberry replied:
Separately: expanding the "extremal" values as you suggest sounds like a good idea, would you take a note about that?

kshitij12345 replied:
> Separately: expanding the "extremal" values as you suggest sounds like a good idea, would you take a note about that?

Already done :)

https://github.com/pytorch/pytorch/blob/e34f285e586809eb0e7a2f90d348c578fe7a0de7/test/test_unary_ufuncs.py#L183-L184

Tested locally and the tests passed. Waiting for green signal on CI

kshitij12345 replied:
> I think I'm missing something here: shouldn't we be randomly generating values within that range as part of test_reference_numerics_hard?

Right now for `hard`
We check only for the following values and not randomly sampled,
https://github.com/pytorch/pytorch/blob/71f4c5c1f436258adc303b710efb3f41b2d50c4e/test/test_unary_ufuncs.py#L59-L62
https://github.com/pytorch/pytorch/blob/71f4c5c1f436258adc303b710efb3f41b2d50c4e/test/test_unary_ufuncs.py#L154-L165

mruberry replied:
Yep; that's makes sense. We should extend make_tensor() to compute values beyond -9 to 9 and change the "hard" test to use a wider range.

kshitij12345 replied:
Noted. Will follow-up on this and extremal extension using finfo in a separate PR.

mruberry commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -3327,6 +3327,44 @@ std::tuple<Tensor, Tensor> polar_backward(
+    // So we manually update gradient for x=0
+    auto eps = std::numeric_limits<scalar_t>::epsilon();
+    auto self_is_not_tiny = self.abs() > eps;
+    auto safe_self =
```
Why do we need to perform the `where` here? I understand the comment but it seems to relate to the later `where` statement

kshitij12345 replied:
Related to this: https://github.com/pytorch/pytorch/issues/52248

where computes the gradient even when that value wasn't actually used. So without replacing the zeroes with `eps` leads to `nan` in the gradient computation.

mruberry replied:
Oh; that's frustrating. Would you add a comment to that effect in the code so we can update this when that bug is fixed?

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -3006,6 +3025,30 @@ def sample_inputs_entr(op_info, device, dtype, requires_grad, **kwargs):
                                     requires_grad=requires_grad)))
 
+
+def sample_inputs_i0_i1(op_info, device, dtype, requires_grad, **kwargs):
```
I'm not really sure what's going on here

The comment suggests that i0e is not differentiable at 0, which is OK, but what about expanding the current unary elementwise default to take a tensor generation function and extend make_tensor to take `exclude_zeros` or `exclude=(0,)` so a partial of it can be passed?

kshitij12345 replied:
So right now, AFAIK, `make_tensor` doesn't guarantee that there would be a zero in the output. `exclude` will just guarantee that there wouldn't be zero. So for others `i0`, `i1` `i1e`, we would still need to manually add `zeros` to the tensor. 
So I think it is ok as is for now.

And I think @krshrimali is adding `exclude` functionality in https://github.com/pytorch/pytorch/pull/57941

mruberry replied:
Yep, #57941 adds exclusions to make_tensor(). 

I see, so this wants no zeros if it's i0e, but does want a zero if it's i0, i1, and i1e? 

I guess the more perfect implementation of that pattern would be that we update the unary sample inputs to accept a make_tensor-like function, so i0e could partial make_tensor(..., exclude_zero=True), but that i0, i1, and i1e want to augment the default unary sample inputs with additional samples, including one with zero.

Is that accurate?

kshitij12345 replied:
Yes. Sounds right.

mruberry replied:
OK; would you do me a favor and just add a TODO comment with that thinking above this function? There may be an opportunity to clarify the comment to reflect that thinking, too

mruberry reviewed (commented):
Hey @kshitij12345! Thanks for being patient around the branch cut.

Overall this looks great; I made some comments inline, and I'm looking forward to hearing your thoughts!

In general, new special operations to float and double for their computation seems good. If users want float16 or bfloat16 support that'd be a good datapoint to hear, and we may need to think more carefully about how to provide those datatypes

kshitij12345 committed:
Merge branch 'master' into develop/special/i1-i1e

kshitij12345 committed:
Merge branch 'develop/special/i1-i1e' into develop/special/i1-i1e

kshitij12345 committed:
remove pseeudo-vectorized path

kshitij12345 committed:
move finfo based values to reference_numerics_hard

kshitij12345 commented:
> I think I'm missing something here: shouldn't we be randomly generating values within that range as part of test_reference_numerics_hard?

Right now for `hard`
We check only for the following values and not randomly sampled,
https://github.com/pytorch/pytorch/blob/71f4c5c1f436258adc303b710efb3f41b2d50c4e/test/test_unary_ufuncs.py#L59-L62
https://github.com/pytorch/pytorch/blob/71f4c5c1f436258adc303b710efb3f41b2d50c4e/test/test_unary_ufuncs.py#L154-L165

kshitij12345 committed:
undo finfo to extremal change

kshitij12345 committed:
add comment for seeminglly unnecessary where

kshitij12345 committed:
add comment for sample_inputs_func

kshitij12345 commented:
@mruberry gentle ping :) this is ready for another round. Have a look whenever free.

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -3138,6 +3157,32 @@ def sample_inputs_entr(op_info, device, dtype, requires_grad, **kwargs):
+#       For more context: https://github.com/pytorch/pytorch/pull/56352#discussion_r633277617
+def sample_inputs_i0_i1(op_info, device, dtype, requires_grad, **kwargs):
+
+    samples = (SampleInput(make_tensor((S,), device, dtype,
```
If the op is torch.special.i0e, doesn't this and the next sample input need to exclude zero?

kshitij12345 replied:
Ideally yes. But at this point we don't have `exclude_zero` yet. Temporarily I will add `x[x == 0] = eps` and add a comment to replace it with `exclude_zero` once that is in. Is that ok?

mruberry replied:
Great idea

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -3138,6 +3157,32 @@ def sample_inputs_entr(op_info, device, dtype, requires_grad, **kwargs):
+                        requires_grad=requires_grad)
+
+        with torch.no_grad():
+            t[torch.randn_like(t) > 0] = 0
```
This isn't guaranteed to set any values to zero. Instead, this could just set the first value of t to zero to ensure one is in the tensor.

mruberry reviewed (commented):
Hey @kshitij12345!

This looks really good overall but the sample inputs still needs a little refinement, I think

kshitij12345 committed:
use astype and explicitly set value to zero in samples

kshitij12345 committed:
handle 0 values for i0e

kshitij12345 committed:
Merge branch 'master' into develop/special/i1-i1e

kshitij12345 committed:
make mypy happy

mruberry reviewed (approved):
Cool! Nice work @kshitij12345. Thanks for persistently pinging me on this one!

kshitij12345 committed:
Merge branch 'master' into develop/special/i1-i1e
